workers/loc.api/sync/positions.snapshot/index.js

```diff
@@ -537,6 +544,150 @@ class PositionsSnapshot {
     }
   }
 
+  _calcPlFromPositionsSnapshots (positionsHistory) {
+    return (
+      positionsSnapshots = [],
+      args = {}
+    ) => {
+      const { mts, timeframe } = args
+
+      // Need to filter duplicate and closed positions as it can be for
+      // week and month and year timeframe in daily positions snapshots
+      // if daily timeframe no need to filter it
+      const positions = this._filterPositionsSnapshots(
+        positionsSnapshots,
+        positionsHistory,
+        timeframe,
+        mts
+      )
+
+      return positions.reduce((accum, curr) => {
+        const { plUsd } = { ...curr }
+        const symb = 'USD'
+
+        if (!Number.isFinite(plUsd)) {
+          return accum
+        }
+
+        return {
+          ...accum,
+          [symb]: Number.isFinite(accum[symb])
+            ? accum[symb] + plUsd
+            : plUsd
+        }
+      }, {})
+    }
+  }
+
+  _filterPositionsSnapshots (
+    positionsSnapshots,
+    positionsHistory,
+    timeframe,
+    mts
+  ) {
+    if (
+      !Array.isArray(positionsSnapshots) ||
+      positionsSnapshots.length === 0
+    ) {
+      return positionsSnapshots
+    }
+
+    return positionsSnapshots.reduce((accum, position) => {
+      if (
+        Number.isFinite(position?.id) &&
+        accum.every((item) => item?.id !== position?.id) &&
+        (
+          timeframe === 'day' ||
+          !this._isClosedPosition(positionsHistory, mts, position?.id)
+        )
+      ) {
+        accum.push(position)
+      }
+
+      return accum
+    }, [])
+  }
+
+  _isClosedPosition (positionsHistory, mts, id) {
+    return (
+      Array.isArray(positionsHistory) &&
+      positionsHistory.length > 0 &&
+      positionsHistory.some((item) => (
+        item.id === id &&
+        item.mts === mts
+      ))
+    )
+  }
+
+  async getPLSnapshot ({
+    auth = {},
+    params = {}
+  } = {}) {
+    const user = await this.authenticator
+      .verifyRequestUser({ auth })
+
+    const {
+      timeframe = 'day',
+      start = 0,
+      end = Date.now()
+    } = params ?? {}
+    const args = {
+      auth: user,
+      params: {
+        timeframe,
+        start,
+        end
+      }
+    }
+
+    const dailyPositionsSnapshotsPromise = this
+      .getSyncedPositionsSnapshot(args)
+    const positionsHistoryPromise = this.dao.getElemsInCollBy(
+      this.ALLOWED_COLLS.POSITIONS_HISTORY,
+      {
+        filter: {
+          user_id: user._id,
```

Comment:
Would this take into consideration the accounts with sub accounts?

Reply:
sure, as sub-account has own _id